kmir/k-src/mir-rvalue.md

```diff
@@ -0,0 +1,371 @@
+
+  syntax BinaryOpName ::= "Add"    [token]
+                        | "Sub"    [token]
+                        | "Mul"    [token]
+                        | "Div"    [token]
+                        | "Rem"    [token]
+                        | "BitXor" [token]
+                        | "BitAnd" [token]
+                        | "BitOr"  [token]
+                        | "Shl"    [token]
+                        | "Shr"    [token]
+                        | "Eq"     [token]
+                        | "Lt"     [token]
+                        | "Le"     [token]
+                        | "Ne"     [token]
+                        | "Ge"     [token]
+                        | "Gt"     [token]
+                        | "Offset" [token]
+
+  syntax Bool ::= isUnOp(String) [function, total]
+  //----------------------------------------------
+  rule isUnOp(OP_NAME)  => true
+    requires OP_NAME ==String "Not"
+      orBool OP_NAME ==String "Neg"
+  rule isUnOp(_)       => false [owise]
+
+  syntax Bool ::= isBinOp(String) [function, total]
+  //-----------------------------------------------
+  rule isBinOp(OP_NAME)  => true
+    requires OP_NAME ==String "Add"
+      orBool OP_NAME ==String "Sub"
+      orBool OP_NAME ==String "Mul"
+      orBool OP_NAME ==String "Div"
+      orBool OP_NAME ==String "Rem"
+      orBool OP_NAME ==String "BitXor"
+      orBool OP_NAME ==String "BitAnd"
+      orBool OP_NAME ==String "BitOr"
+      orBool OP_NAME ==String "Shl"
+      orBool OP_NAME ==String "Shr"
+      orBool OP_NAME ==String "Eq"
+      orBool OP_NAME ==String "Lt"
+      orBool OP_NAME ==String "Le"
+      orBool OP_NAME ==String "Ne"
+      orBool OP_NAME ==String "Ge"
+      orBool OP_NAME ==String "Gt"
+      orBool OP_NAME ==String "Offset"
+  rule isBinOp(_)       => false [owise]
+
+  syntax CheckedBinaryOpName ::= "CheckedAdd"
+                               | "CheckedSub"
+                               | "CheckedMul"
+                               | "CheckedShl"
+                               | "CheckedShr"
+
+endmodule
+```
+
+Evaluation of rvalues
+---------------------
+
+```k
+module MIR-RVALUE
+  imports MIR-RVALUE-SYNTAX
+  imports MIR-BUILTINS-SYNTAX
+  imports MIR-TYPES
+  imports MIR-CONFIGURATION
+```
+
+Evaluate a syntactic `RValue` into a semantics `RValueResult`. Inspired by [eval_rvalue_into_place](https://github.com/rust-lang/rust/blob/bd43458d4c2a01af55f7032f7c47d7c8fecfe560/compiler/rustc_const_eval/src/interpret/step.rs#L148).
+
+```k
+  syntax InterpResult ::= evalRValue(FunctionLikeKey, RValue) [function]
+  //--------------------------------------------------------------------
+  rule evalRValue(FN_KEY, VALUE:Operand)   => evalOperand(FN_KEY, VALUE)
+  rule evalRValue(FN_KEY, UN_OP:UnaryOp)   => evalUnaryOp(FN_KEY, UN_OP)
+  rule evalRValue(FN_KEY, BIN_OP:BinaryOp) => evalBinaryOp(FN_KEY, BIN_OP)
+  rule evalRValue(_FN_KEY, RVALUE)         => Unsupported(RVALUE) [owise]
+```
+
+### `Operand` evaluation
```

Comment:
Coud the evaluation of `Operand` be moved together with the definition of `Operand`?

Reply:
The syntax definition of the `Operand` sort (defined in `MIR-RVALUE-SYNTAX` module) and it's evaluations semantics (defined in `MIR-RVALUE` module) are` separated in purpose. The `*-SYNTAX` modules are used to produce the Bison parser, and thus should not import the semantics modules. Perhaps there's a way to keep them together, but, to the best of my knowledge, this is the preferred way to structure K definitions.